config/soap-client-notification.php

```diff
@@ -0,0 +1,123 @@
+    (new OpenEuropa\Assembler\NullableGetterAssemblerOptions())
+        ->blacklist($specialClassesAndProperties)
+);
+
+$fluentAdderAssembler = new OpenEuropa\Assembler\FluentAdderAssembler(
+    (new OpenEuropa\Assembler\FluentAdderAssemblerOptions())
+        ->whitelist($specialClassesAndProperties)
+);
+
+$hasPropertyAssembler = new OpenEuropa\Assembler\HasPropertyAssembler();
+
+return Config::create()
+    ->setWsdl('resources/NotificationServiceWSDL.xml')
+    ->setTypeDestination('src/Notification/Type')
+    ->setTypeNamespace('OpenEuropa\EPoetry\Notification\Type')
+    ->setClientDestination('src/Notification/')
+    ->setClientName('NotificationClient')
+    ->setClientNamespace('OpenEuropa\Notification')
+    ->setClassMapDestination('src/Notification/')
+    ->setClassMapName('NotificationClassmap')
+    ->setClassMapNamespace('OpenEuropa\EPoetry\Notification')
+// We have to do this as the SOAP handler will erroneously create duplicate
+// public properties when a value object extends another one with those
+// same properties marked as "private".
+    ->setRuleSet(
+        new Rules\RuleSet(
+            [
+                new Rules\AssembleRule(new Assembler\ClassMapAssembler()),
+            ]
+        )
+    )
+    ->addRule(
+        new Rules\TypenameMatchesRule(
+            new Rules\AssembleRule(new Assembler\ResultAssembler()),
+            '/Response$/'
+        )
+    )
+// Generate SOAP request classes.
+//
+// Request objects must implement \Phpro\SoapClient\Type\RequestInterface
+// The rule matches the following SOAP types:
+//
+// - receiveNotification
+    ->addRule(
+        new Rules\TypenameMatchesRule(
+            new Rules\AssembleRule(new Assembler\RequestAssembler()),
+            '/(ReceiveNotification)$/'
+        )
+    )
+// Set the default property assembler and generate all properties.
+    ->addRule(new Rules\AssembleRule($defaultPropertyAssembler))
+// Update properties and set them as 'nullable'
+    ->addRule(new Rules\AssembleRule($arrayPropertyAssembler))
+// Set the default setter assembler and generate all setters methods.
+    ->addRule(new Rules\AssembleRule($nullablePropertyAssembler))
+// Update properties and update only some of them.
+    ->addRule(new Rules\AssembleRule($defaultSetterAssembler))
+// Update setters and update only some of them.
+    ->addRule(new Rules\AssembleRule($arraySetterAssembler))
+// Set the default getter assembler and generate all getters methods.
+    ->addRule(new Rules\AssembleRule($defaultGetterAssembler))
+// Update getters and set them as 'nullable'
+    ->addRule(new Rules\AssembleRule($nullableGetterAssembler))
+// Update getters and update only some of them.
+    ->addRule(new Rules\AssembleRule($arrayGetterAssembler))
+// Add adders only on some classes only.
+    ->addRule(new Rules\AssembleRule($fluentAdderAssembler))
+// Add has[Properties] only on some classes only.
+    ->addRule(new Rules\AssembleRule($hasPropertyAssembler));
```

Comment:
Can we collect all these rules into a single factory class that both notification and request will share? So we maintain our rules collection in one single place. This will also help when porting to Soap Client version 1.0.0.

I'd propose the following:

- Create a class: `OpenEuropa\EPoetry\CodeGenerator\ConfigFactory`
- Move all the shareable code above, including comments etc. in a static method of that class, like:

```php
class ConfigFactory {

    /**
     * @param array $specialClassesAndProperties
     *   Mapping of classes and properties that will receive a special treatment
     *   in the assemblers. Sometimes they are used as a whitelist, sometimes as
     *   blacklist.
     *
     * @return \Phpro\SoapClient\CodeGenerator\Config\Config
     *   Configuration object instance.
     */
    static public function create(array $specialClassesAndProperties): Config
    {
        // Set all property visibility to "protected".
        // We have to do this as the SOAP handler will erroneously create duplicate
        // public properties when a value object extends another one with those
        // same properties marked as "private".
        $defaultPropertyAssembler = new Assembler\PropertyAssembler(PropertyGenerator::VISIBILITY_PROTECTED);

        $arrayPropertyAssembler = new OpenEuropa\Assembler\ArrayPropertyAssembler(
            (new OpenEuropa\Assembler\ArrayPropertyAssemblerOptions())
                ->whitelist($specialClassesAndProperties)
        );

        // ...
        return Config::create()
            ->setRuleSet(
                new Rules\RuleSet(
                    [
                        new Rules\AssembleRule(new Assembler\ClassMapAssembler()),
                    ]
                )
            )
            ->addRule(
                new Rules\TypenameMatchesRule(
                    new Rules\AssembleRule(new Assembler\ResultAssembler()),
                    '/Response$/'
                )
            ),
        // ...
    }
}
```

Then in each file we call that method and add specific configurations, like:

```php
return ConfigFactory::create($specialClassesAndProperties)
    ->setWsdl('resources/dgtServiceWSDL.xml')
    ->setTypeDestination('src/Request/Type')
    ->setTypeNamespace('OpenEuropa\EPoetry\Request\Type')
    ->setClientDestination('src')
    ->setClientName('RequestClient')
    ->setClientNamespace('OpenEuropa\EPoetry\Request')
    ->setClassMapDestination('src/Request')
    ->setClassMapName('RequestClassmap')
    ->setClassMapNamespace('OpenEuropa\EPoetry\Request');
```

Reply:
Rules are different for each case so they cannot be stored in a unique place.
Anyway, I suggest this improvement be done in a followup. It's a nice to have.

Reply:
Good idea. Done.